File: TinkoffChat/Presentation Layer/Modules/ThemesModule/ThemesScene/View/ThemesViewController.m
//
//  ThemesViewController.m
//  TinkoffChat
//
//  Created by st.i on 13/10/2018.
//  Copyright © 2018 Иван Стефанов. All rights reserved.
//

#import "ThemesViewController.h"
#import "ThemesViewControllerDelegate.h"

/*
 Чтобы использовать VC на Objective-C:
 1. Поставить галочку в таргете
 2. Раскомментировать код для themesDelegate в ConversationsListViewController
 3. Убрать галочку "Inherit module from target" в ThemesStoryboard
 4. Раскомментировать #import "ThemesViewController.h" в Bridging-Header
 */

@interface ThemesViewController ()

@end

@implementation ThemesViewController 

- (void)viewDidLoad {
    [super viewDidLoad];
    
    UIColor *theme1 = [UIColor whiteColor];
    UIColor *theme2 = [UIColor darkGrayColor];
    UIColor *theme3 = [UIColor yellowColor];
//    _model = [[Themes alloc] init];
    _model.theme1 = theme1;
    _model.theme2 = theme2;
    _model.theme3 = theme3;    
}

- (void)viewWillAppear:(BOOL)animated {
    [super viewWillAppear:animated];
    
    NSData *appTheme = [[NSUserDefaults standardUserDefaults]objectForKey:@"appTheme"];
    if (appTheme != nil) {
        UIColor *appThemeColor = [NSKeyedUnarchiver unarchivedObjectOfClass:UIColor.class fromData:appTheme error:nil];
        self.view.backgroundColor = appThemeColor;
    }
}

- (void)dealloc {
//    [_model release];
    _model = nil;
    _themesDelegate = nil;
//    [super dealloc];
}

#pragma mark - Getters

- (id<​ThemesViewControllerDelegate>)themesDelegate {
    return _themesDelegate;
}

- (Themes *)model {
    return _model;
}

#pragma mark - Setters

- (void)setThemesDelegate:(id<​ThemesViewControllerDelegate>)themesDelegate {
    if (_themesDelegate != themesDelegate) {
        _themesDelegate = themesDelegate;
    }
}

- (void)setModel:(Themes *)model {
    if (_model != model) {
//        [_model release];
//        _model = [model retain];
        _model = model;
    }
}

#pragma mark - Actions

- (IBAction)closeViewController:(UIBarButtonItem *)sender {
    [self dismissViewControllerAnimated:YES completion:nil];
}

- (IBAction)changeColorTheme:(UIButton *)sender {
    UIColor *selectedTheme;
    switch (sender.tag) {
        case 1:
            selectedTheme = _model.theme1;
            break;
        case 2:
            selectedTheme = _model.theme2;
            break;
        case 3:
            selectedTheme = _model.theme3;
            break;
        default:
            selectedTheme = [UIColor whiteColor];
            break;
    }
    if ([_themesDelegate respondsToSelector:@selector(themesViewController:didSelectTheme:)]) {
        [_themesDelegate themesViewController:self didSelectTheme:selectedTheme];
    }
}

@end

File: TinkoffChat/Presentation Layer/Modules/ThemesModule/ThemesSwiftScene/View/ThemesViewController.swift
//
//  ThemesViewController.swift
//  TinkoffChat
//
//  Created by st.i on 16/10/2018.
//  Copyright © 2018 Иван Стефанов. All rights reserved.
//

import UIKit

/*
 Чтобы использовать VC на Swift'е:
 1. Поставить галочку в таргете
 2. Раскомментировать код для closure в ConversationsListViewController
 3. Поставить галочку "Inherit module from target" в ThemesStoryboard
 4. Закомментировать #import "ThemesViewController.h" в Bridging-Header
 */

class ThemesViewController: UIViewController {
    
    var didSelectThemeClosure: ((_ selectedTheme: UIColor?) -> Void)?
    var model: Themes?

    override func viewDidLoad() {
        super.viewDidLoad()

        let theme1: UIColor = .white
        let theme2: UIColor = .darkGray
        let theme3: UIColor = .yellow
        
        model = Themes()
        model?.theme1 = theme1
        model?.theme2 = theme2
        model?.theme3 = theme3
    }
    
    override func viewWillAppear(_ animated: Bool) {
        super.viewWillAppear(animated)
        
        if let savedColorData = UserDefaults.standard.data(forKey: "appTheme") {
            do {
                let savedColor = try NSKeyedUnarchiver.unarchivedObject(ofClass: UIColor.self, from: savedColorData)
                view.backgroundColor = savedColor
            } catch {
                print("Can't unarchive app theme")
            }
        }
    }

    @IBAction func closeViewController(_ sender: UIBarButtonItem) {
        dismiss(animated: true, completion: nil)
    }
    
    @IBAction func changeColorTheme(_ sender: UIButton) {
        var selectedTheme: UIColor?
        switch (sender.tag) {
        case 1:
            selectedTheme = model?.theme1
            break
        case 2:
            selectedTheme = model?.theme2
            break
        case 3:
            selectedTheme = model?.theme3
            break
        default:
            selectedTheme = .white
            break
        }
        didSelectThemeClosure!(selectedTheme)
    }
}
